src/SDAO.t.sol

```diff
@@ -0,0 +1,509 @@
+
+        address owner = vm.addr(privateKey);
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+            privateKey,
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    token.DOMAIN_SEPARATOR(),
+                    keccak256(abi.encode(PERMIT_TYPEHASH, owner, to, amount, nonce, deadline))
+                )
+            )
+        );
+
+        vm.expectRevert("SDAO/invalid-permit");
+        token.permit(owner, to, amount, deadline, v, r, s);
+    }
+
+    function testRevertPermitBadDeadlineFuzz(uint248 privateKey, address to, uint256 amount, uint256 deadline) public {
+        deadline = bound(deadline, block.timestamp, block.timestamp + type(uint80).max);
+        privateKey = uint248(bound(privateKey, 1, type(uint248).max));
+
+        address owner = vm.addr(privateKey);
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+            privateKey,
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    token.DOMAIN_SEPARATOR(),
+                    keccak256(abi.encode(PERMIT_TYPEHASH, owner, to, amount, 0, deadline))
+                )
+            )
+        );
+
+        vm.expectRevert("SDAO/invalid-permit");
+        token.permit(owner, to, amount, deadline + 1, v, r, s);
+    }
+
+    function testRevertPermitPastDeadlineFuzz(uint248 privateKey, address to, uint256 amount, uint256 deadline) public {
+        deadline = bound(deadline, 0, block.timestamp - 1);
+        privateKey = uint248(bound(privateKey, 1, type(uint248).max));
+
+        address owner = vm.addr(privateKey);
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+            privateKey,
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    token.DOMAIN_SEPARATOR(),
+                    keccak256(abi.encode(PERMIT_TYPEHASH, owner, to, amount, 0, deadline))
+                )
+            )
+        );
+
+        vm.expectRevert("SDAO/permit-expired");
+        token.permit(owner, to, amount, deadline, v, r, s);
+    }
+
+    function testRevertPermitReplayFuzz(uint248 privateKey, address to, uint256 amount, uint256 deadline) public {
+        deadline = bound(deadline, block.timestamp, block.timestamp + type(uint80).max);
+        privateKey = uint248(bound(privateKey, 1, type(uint248).max));
+
+        address owner = vm.addr(privateKey);
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+            privateKey,
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    token.DOMAIN_SEPARATOR(),
+                    keccak256(abi.encode(PERMIT_TYPEHASH, owner, to, amount, 0, deadline))
+                )
+            )
+        );
+        token.permit(owner, to, amount, deadline, v, r, s);
+
+        vm.expectRevert("SDAO/invalid-permit");
+        token.permit(owner, to, amount, deadline, v, r, s);
+    }
```

Comment:
Would also include the `checkAuth()` and `checkModifiers()` tests to verify both the auth stuff is implemented properly and that the `mint()` function has the `auth` modifier.

Reply:
Done